modules/javafx.graphics/src/main/native-glass/gtk/wrapped.c

```diff
@@ -148,3 +148,71 @@ void wrapped_gdk_x11_display_set_window_scale (GdkDisplay *display,
     }
 }
 
+static GdkSeat * (*_gdk_display_get_default_seat) (GdkDisplay *display);
+
+// Note added in libgdk 3.20 which is > our OEL 7.0 version of 3.8
+GdkSeat * wrapped_gdk_display_get_default_seat (GdkDisplay *display)
+{
+    if (_gdk_display_get_default_seat == NULL) {
+        _gdk_display_get_default_seat = dlsym(RTLD_DEFAULT, "gdk_display_get_default_seat");
+        if (gtk_verbose && _gdk_display_get_default_seat) {
+            fprintf(stderr, "loaded gdk_display_get_default_seat\n"); fflush(stderr);
+        }
+    }
+
+    if (_gdk_display_get_default_seat != NULL) {
+        return (*_gdk_display_get_default_seat)(display);
+    }
+
+    return NULL;
+}
+
+static GdkGrabStatus (*_gdk_seat_grab) (GdkSeat* seat, GdkWindow* window,
+                            int capabilities /* GdkSeatCapabilities capabilities */,
+                            gboolean owner_events, GdkCursor* cursor, const GdkEvent* event,
+                            void * prepare_func /* GdkSeatGrabPrepareFunc prepare_func */,
+                            gpointer prepare_func_data);
+
+// Note added in libgdk 3.20 which is > our OEL 7.0 version of 3.8
+gboolean wrapped_gdk_seat_grab(GdkWindow* window,
+                            int capabilities /* GdkSeatCapabilities capabilities */,
+                            gboolean owner_events, GdkCursor* cursor, const GdkEvent* event,
+                            void * prepare_func /* GdkSeatGrabPrepareFunc prepare_func */,
+                            gpointer prepare_func_data,
+                            GdkGrabStatus* status)
+{
+    if (_gdk_seat_grab == NULL) {
+        _gdk_seat_grab = dlsym(RTLD_DEFAULT, "gdk_seat_grab");
+        if (gtk_verbose && _gdk_seat_grab) {
+            fprintf(stderr, "loaded gdk_seat_grab\n"); fflush(stderr);
+        }
+    }
+
+    GdkSeat* seat = wrapped_gdk_display_get_default_seat(gdk_window_get_display(window));
+    if (seat != NULL && _gdk_seat_grab != NULL) {
+        *status = (*_gdk_seat_grab)(seat, window, capabilities,
+                                 owner_events, cursor, event, prepare_func, prepare_func_data);
+        return TRUE;
+    }
+    return FALSE;
```

Comment:
I did try to test on Ubuntu 16.04 and compilation failed (no surprise because `GdkSeat` does not exists there). Suggestion to keep  `#ifdef` here and `return FALSE` on `#else` so it would still compile on Ubuntu 16.04 and older systems. Will need to `#ifdef` all `GdkSeat` usage.

Reply:
I take `GdkSeat` is available since GTK 3.0? https://docs.gtk.org/gdk3/class.Seat.html

But don't we have a minimum set on 3.8.0?

Would this work?
```
#if GTK_CHECK_VERSION(3, 0, 0)
static GdkSeat * (*_gdk_display_get_default_seat) (GdkDisplay *display);
GdkSeat * wrapped_gdk_display_get_default_seat (GdkDisplay *display)
{...}
#endif

...

#if GTK_CHECK_VERSION(3, 0, 0)
    GdkSeat* seat = wrapped_gdk_display_get_default_seat(gdk_window_get_display(window));
    if (seat != NULL && _gdk_seat_grab != NULL) {
        *status = ...
        return TRUE;
    }
#endif
```

Reply:
You're right. GdkSeat should be available. Maybe It needs an include?

Reply:
I think the docs are wrong, I probably exists since 3.20, so maybe check for `GTK_CHECK_VERSION(3, 20, 0);`.

`/home/tsayao/jose/jfx/modules/javafx.graphics/src/main/native-glass/gtk/wrapped.c:200:34: error: unknown type name ‘GdkSeat’
`

Reply:
Okay, that is unfortunate (GTK docs inaccurate), but makes sense.

I'll add the compile-time checks to `wrapped.c`.

Reply:
The check is to allow compilation and test on Ubuntu 16.04 - When shipping the final build it should be built on gtk 3.20+, so both dlsym and the check makes sense. 

So when running on Ubuntu 16.04:
- For testing on the platform, the #ifdef will make it possible to compile;
- For running the released built (which should be built on gtk 3.20+) dlsym will fail and fallback;

Reply:
That's only if we want to keep building working on 16.04. I think it makes easier to test on it.
But, it's already failing for Platform preferences API code.

Reply:
In any case, if `GdkSeat` is available only since 3.20, then we need to add the compile-time checks anyway, since minimum supported is 3.8.

Reply:
It seems like this is the best we can do for now. What is means is that we won't be able to build on a system with 3.8 - 3.19 and distribute a library that will use the new functionality when running on 3.20 or later. That's probably OK in this case, but is worth noting.

What is important, and why the runtime check is needed (as noted above) is that we can build a production release of JavaFX on a system that has a later GTK and it will run on a system with an older GTK. It won't use the feature on those older systems, but won't crash.

Reply:
As an FYI, we use GTK 3.22 on our CI build machines so I wouldn't want to see the build-time dependency move any higher than it currently is.

Reply:
The suggestion was to allow building on 16.04 - so it's easier to test. But it currently does not build because of code introduced on Platform Preferences API. On the other hand, maybe it's best to fail so we make sure building requires 3.20.

Reply:
The native Platform Preferences code is failing for an entirely different reason (a too-old compiler), but you raise an interesting point.

If we really can't support building a distributable binary on a system with a too-old GTK library in a way that enables using a new feature like this, we might prefer to "fail fast" at build time, rather than produce a binary that doesn't support the new capability even when running on a newer system.

If we decide that is a reasonable thing to do, we might want the minimum _build-time_ version of GTK to be higher than the minimum version we support at _runtime_.

I think we shouldn't necessarily tie that decision to this PR. So I think the current approach in this PR: allow it to build on an older GTK without using the new functionality, is probably OK (although not ideal as evidenced by this discussion).